src/plonk/mod.rs

```diff
@@ -478,115 +554,203 @@ impl<F: PrimeField> TableData<F> {
+            round_sizes.push(2 * num_lookups * nrow);
+        } else if num_lookups > 0 {
+            // advice columns || (l_i, m_i)
+            round_sizes.push((self.cs.num_advice_columns() + 2 * num_lookups) * nrow);
+            // (h_i, g_i)
+            round_sizes.push(2 * num_lookups * nrow);
         } else {
-            Expression::from_halo2_expr(
-                &self.cs.gates()[0].polynomials()[0],
-                self.cs.num_selectors(),
-                self.cs.num_fixed_columns(),
-            )
-            .expand()
+            // advice columns
+            round_sizes.push(self.cs.num_advice_columns() * nrow);
         };
 
+        let exprs = self
+            .cs
+            .gates()
+            .iter()
+            .flat_map(|gate| gate.polynomials().iter().cloned())
+            .map(|expr| {
+                Expression::from_halo2_expr(
+                    &expr,
+                    self.cs.num_selectors(),
+                    self.cs.num_fixed_columns(),
+                    self.num_lookups(),
+                )
+            })
+            .chain(self.lookup_exprs(&self.cs))
+            .collect::<Vec<_>>();
+
+        // we use r3 to combine all custom gates and lookup expressions
+        // find the challenge index of r3
+        let cha_index = if num_challenges > 0 {
+            num_challenges - 1
+        } else {
+            0
+        };
+        let poly = compress_expression(&exprs[..], cha_index).expand();
         let permutation_matrix = self.permutation_matrix();
 
         PlonkStructure {
             k: self.k as usize,
             selectors,
-            fixed_columns,
+            fixed_columns: self.fixed_columns.clone(),
             num_advice_columns: self.cs.num_advice_columns(),
             num_challenges,
-            gate,
+            round_sizes,
+            poly,
             fixed_commitment,
             permutation_matrix,
-            lookups: self.lookups.clone(),
+            lookup_arguments: self.lookup_arguments.clone(),
         }
     }
 
-    pub fn plonk_instance<C: CurveAffine<ScalarExt = F>>(
+    /// run 0-round special soundness protocol
+    /// i.e no need to generate challenge
+    pub fn run_sps_protocol_0<C: CurveAffine<ScalarExt = F>>(
         &self,
         ck: &CommitmentKey<C>,
-    ) -> PlonkInstance<C> {
-        let W = self.plonk_witness().W;
-        let W_commitment = ck.commit(&W[..]);
-        let mut instance: Vec<C::ScalarExt> = Vec::new();
-        for inst in self.instance.iter() {
-            instance.push(*inst)
-        }
-        PlonkInstance {
-            W_commitment,
-            instance,
-            challenges: vec![],
-        }
+    ) -> (PlonkInstance<C>, PlonkWitness<F>) {
+        assert!(
+            !self.advice.is_empty(),
+            "should call TableData.assembly() first"
+        );
+
+        let W1 = concatenate_with_padding(&self.advice_columns[..], 2usize.pow(self.k));
```

Comment:
You almost never need `[..]`

Reply:
yeah, sometimes will forget, bad habit...